### PR TITLE
Add how to filter binlog events using SQL expressions

### PR DESCRIPTION
**Pull request by TomShawn**
### What is changed, added or deleted? (Required)
Add how to filter binlog events using SQL expressions

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions](https://github.com/pingcap/docs/blob/master/CONTRIBUTING.md#guideline-for-choosing-the-affected-versions).

- [ ] master (the latest development version)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from: data-migration/migrate-with-binlog-sql-expression-filter.md in https://github.com/pingcap/docs-cn/pull/7480
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: data-migration/migrate-with-binlog-sql-expression-filter.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+---
+
+# Filter Binlog Events Using SQL Expressions
+
+This document introduces how to filter binlog events using SQL expressions when you use DM to perform continuous incremental data replication. For the detailed replication instruction, refer to the following documents:
+
+- [Migrate MySQL data of less than 1 TB to TiDB](/data-migration/migrate-mysql-tidb-less-tb.md)
+- [Migrate MySQL data of more than 1 TB to TiDB](/data-migration/migrate-mysql-tidb-above-tb.md)
+- [Merge and migrate MySQL shards of less than 1 TB to TiDB](/data-migration/migrate-shared-mysql-tidb-less-tb.md)
+- [Merge and migrate MySQL shards of more than 1 TB to TiDB](/data-migration/migrate-shared-mysql-tidb-above-tb.md)
```
Comment (en-jin19) [inline data-migration/migrate-with-binlog-sql-expression-filter.md:13]:
```suggestion
- [Merge and migrate MySQL shards of less than 1 TB to TiDB](/data-migration/migrate-sharding-mysql-tidb-less-tb.md)
- [Merge and migrate MySQL shards of more than 1 TB to TiDB](/data-migration/migrate-sharding-mysql-tidb-above-tb.md)
```

File: data-migration/migrate-with-binlog-sql-expression-filter.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+When performing incremental data replication, you can use the [Binlog Event Filter](/data-migration/migrate-with-binlog-event-filter.md) to filter certain types of binlog events, for example, not to replicate `DELETE` events to the downstream for archiving and audit purposes. However, the Binlog Event Filter cannot determine whether to filter the `DELETE` event of a row that requires finer granularity.
```
Comment (en-jin19) [inline data-migration/migrate-with-binlog-sql-expression-filter.md:15]:
```suggestion
When performing incremental data replication, you can use the [Binlog Event Filter](/data-migration/migrate-with-binlog-event-filter.md) to filter certain types of binlog events. For example, you can choose not to replicate `DELETE` events to downstream for the purposes like archiving and auditing. However, the Binlog Event Filter cannot determine whether to filter the `DELETE` event of a row that requires finer granularity.
```

File: data-migration/migrate-with-binlog-sql-expression-filter.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+To address the issue, since v2.0.5, DM supports using `binlog value filter` in incremental data replication to filter data. Among the DM-supported and `ROW`-formatted binlog, the binlog events carry values of all columns, and you can configure SQL expressions based on these values. If the expression calculates a row change as `TRUE`, DM does not replicate this row change to the downstream.
+
+Similar to [Binlog Event Filter](/data-migration/migrate-with-binlog-event-filter.md), you need to configure `binlog value filter` in the task configuration file. See the following configuration example for details. For the advanced task configuration and the description, refer to [DM advanced task configuration file](https://docs.pingcap.com/tidb-data-migration/stable/task-configuration-file-full#task-configuration-file-template-advanced).
```
Comment (en-jin19) [inline data-migration/migrate-with-binlog-sql-expression-filter.md:19]:
```suggestion
Similar to [Binlog Event Filter](/data-migration/migrate-with-binlog-event-filter.md), you need to configure `binlog value filter` in the task configuration file. For details, see the following configuration example. For the advanced task configuration and the description, refer to [DM advanced task configuration file](https://docs.pingcap.com/tidb-data-migration/stable/task-configuration-file-full#task-configuration-file-template-advanced).
```

File: data-migration/migrate-with-binlog-sql-expression-filter.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+task-mode: all
+mysql-instances:
+  - source-id: "mysql-replica-01"
+    expression-filters: ["even_c"]
+expression-filter:
+  even_c:
+    schema: "expr_filter"
+    table: "tbl"
+    insert-value-expr: "c % 2 = 0"
+```
```
Comment (en-jin19) [inline data-migration/migrate-with-binlog-sql-expression-filter.md:32]:
```suggestion
```yaml
name: test
task-mode: all

mysql-instances:
  - source-id: "mysql-replica-01"
    expression-filters: ["even_c"]

expression-filter:
  even_c:
    schema: "expr_filter"
    table: "tbl"
    insert-value-expr: "c % 2 = 0"
```
```

File: data-migration/migrate-with-binlog-sql-expression-filter.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+In the above configuration example, the `even_c` rule is configured and referenced by the data source `mysql-replica-01`. According to this rule, for the `tb1` table in the `expr_filter` schema, when an even number is inserted into the `c` column (`c % 2 = 0`), this `insert` statement is not replicated to the downstream. The following example shows the effect of the this rule.
```
Comment (en-jin19) [inline data-migration/migrate-with-binlog-sql-expression-filter.md:34]:
```suggestion
In the above configuration example, the `even_c` rule is configured and referenced by the data source `mysql-replica-01`. According to this rule, for the `tb1` table in the `expr_filter` schema, when an even number is inserted into the `c` column (`c % 2 = 0`), this `insert` statement is not replicated to the downstream. The following example shows the effect of this rule.
```

File: data-migration/migrate-with-binlog-sql-expression-filter.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
++------+------+
+| id   | c    |
++------+------+
+|    1 |    1 |
+|    3 |    3 |
++------+------+
+2 rows in set (0.001 sec)
+```
+
+## Configuration parameter and description
```
Comment (en-jin19) [inline data-migration/migrate-with-binlog-sql-expression-filter.md:55]:
```suggestion
## Configuration parameters and description
```

File: data-migration/migrate-with-binlog-sql-expression-filter.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+- `delete-value-expr`: Configures an expression that takes effect on values carried by the `DELETE` type binlog events (DELETE_ROWS_EVENT). You cannot use this expression together with `insert-value-expr`, `update-old-value-expr` or `update-new-value-expr`.
+
+> **Note:**
+>
+> - You can configure `update-old-value-expr` and `update-new-value-expr` together.
+> - When `update-old-value-expr` and `update-new-value-expr` are configured together, the rows whose "update + old values" meet `update-old-value-expr` and whose "update + new values" meet `update-new-value-expr` are filtered.
+> - When one of `update-old-value-expr` and `update-new-value-expr` is configured, the configured expression determines whether to filter the entire row change, which means that the deletion of old values and the insertion of new values are filtered as a whole. You can use the SQL expression on one column or on multiple columns. You can also use the SQL functions supported by TiDB, such as `c % 2 = 0`, `a*a + b*b = c*c`, and `ts > NOW()`.
```
Comment (en-jin19) [inline data-migration/migrate-with-binlog-sql-expression-filter.md:68]:
```suggestion
> - When one of `update-old-value-expr` and `update-new-value-expr` is configured, the configured expression determines whether to filter the **entire row change**, which means that the deletion of old values and the insertion of new values are filtered as a whole. 

You can use the SQL expression on one column or on multiple columns. You can also use the SQL functions supported by TiDB, such as `c % 2 = 0`, `a*a + b*b = c*c`, and `ts > NOW()`.
```

File: data-migration/migrate-with-binlog-sql-expression-filter.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+The `TIMESTAMP` default time zone is the time zone specified in the task configuration file. The default value is the time zone of the downstream. You can explicitly specify the time zone in a way like `c_timestamp = '2021-01-01 12:34:56.5678+08:00'`.
+
+You can configure multiple filtering rules under the `expression-filter` configuration item. The upstream data source references the required rule in `expression-filters` to make it effective. When multiple rules are used, if any one of the rules are matched, the entire row change is filtered.
```
Comment (en-jin19) [inline data-migration/migrate-with-binlog-sql-expression-filter.md:72]:
```suggestion
You can configure multiple filtering rules under the `expression-filter` configuration item. The upstream data source references the required rule in `expression-filters` to make it effective. When multiple rules are used, if **any** one of the rules are matched, the entire row change is filtered.
```

File: data-migration/migrate-with-binlog-sql-expression-filter.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+mysql-instances:
+  - source-id: "mysql-replica-01"
+    expression-filters: ["even_c"]
+
+expression-filter:
+  even_c:
+    schema: "expr_filter"
+    table: "tbl"
+    insert-value-expr: "c % 2 = 0"
```
Comment (TomShawn) [inline data-migration/migrate-with-binlog-sql-expression-filter.md:33]:
```suggestion
    insert-value-expr: "c % 2 = 0"
```
```

**Comment by TomShawn** [comment]
@D3Hunter @en-jin19 All comments are addressed. PTAL again. Thanks!

**Review by D3Hunter — APPROVED** [review]
LGTM

File: data-migration/migrate-with-binlog-sql-expression-filter.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+- `update-old-value-expr`: Configures an expression that takes effect on the old values carried by the `UPDATE` type binlog events (UPDATE_ROWS_EVENT). You cannot use this expression together with `insert-value-expr` or `delete-value-expr` in the same configuration item.
+- `update-new-value-expr`: Configures an expression that takes effect on the new values carried by the `UPDATE` type binlog events (UPDATE_ROWS_EVENT). You cannot use this expression together with `insert-value-expr` or `delete-value-expr` in the same configuration item.
+- `delete-value-expr`: Configures an expression that takes effect on values carried by the `DELETE` type binlog events (DELETE_ROWS_EVENT). You cannot use this expression together with `insert-value-expr`, `update-old-value-expr` or `update-new-value-expr`.
+
+> **Note:**
+>
+> - You can configure `update-old-value-expr` and `update-new-value-expr` together.
+> - When `update-old-value-expr` and `update-new-value-expr` are configured together, the rows whose "update + old values" meet `update-old-value-expr` and whose "update + new values" meet `update-new-value-expr` are filtered.
```
Comment (en-jin19) [inline data-migration/migrate-with-binlog-sql-expression-filter.md:69]:
```suggestion
> - When `update-old-value-expr` and `update-new-value-expr` are configured together, the rows whose "update + old values" meet `update-old-value-expr` **and** whose "update + new values" meet `update-new-value-expr` are filtered.
```

**Review by en-jin19 — APPROVED** [review]
LGTM

File: data-migration/migrate-with-binlog-sql-expression-filter.md [hunk 11]
```diff
@@ -0,0 +1,80 @@
+---
+title: Filter Binlog Events Using SQL Expressions
+summary: Learn how to filter binlog events using SQL expressions.
+---
+
+# Filter Binlog Events Using SQL Expressions
```
Comment (TomShawn) [inline data-migration/migrate-with-binlog-sql-expression-filter.md:6]:
```suggestion
title: Filter DML Events Using SQL Expressions
summary: Learn how to filter DML events using SQL expressions.
---

# Filter DML Events Using SQL Expressions
```

**Pull request merged**